File: backtest/cfe_sweep.cpp
// cfe_sweep.cpp — exact C++ translation of cfe_sweep.py
// Grid and logic match Python 1:1. No extra params.
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <deque>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

// ── Params (mirrors Python dict) ─────────────────────────────────────────────
struct P {
    int    rn;          // RSI period (also controls EMA alpha = 2/(rn+1))
    double rt;          // RSI trend threshold (low)
    double rm;          // RSI trend threshold (max)
    double dm;          // drift min
    double st;          // sustained drift threshold
    int64_t sm;         // sustained duration (ms)
    double sl;          // SL = sl * ATR
    double tp;          // TP = tp * SL
    int64_t cd;         // cooldown (ms)
    int64_t mh;         // max hold (ms)
    bool   tr;          // trail enabled
    double ta;          // trail arm (pts)
    double td;          // trail dist (pts)
};

// ── Trade record ─────────────────────────────────────────────────────────────
struct Trade {
    double pnl, mfe;
    bool   il;          // long
    int    h;           // held seconds
    char   r[8];        // reason
};

// ── CFE engine (mirrors Python CFE class exactly) ─────────────────────────────
struct Eng {
    P p;

    // RSI state (mirrors ursi())
    std::deque<double> rg, rl;
    double rpm = 0.0;
    double rc = 50.0, rp = 50.0, rt_val = 0.0;
    bool   rw = false;
    double ra;           // EMA alpha = 2/(rn+1)
    std::deque<double> rm_buf; // recent mid (last 10)

    // drift sustained state
    int    dsd = 0;      // direction: 1, -1, 0
    int64_t dss = 0;     // ms when direction was set

    // position state
    bool   pa = false;   // position active
    bool   pl = false;   // long
    bool   pbe = false;  // breakeven hit
    double pe = 0;       // entry price
    double psl = 0;      // stop loss
    double ptp = 0;      // take profit
    int64_t pts = 0;     // position open ts
    double pm = 0;       // max MFE
    double ps = 0.01;    // size

    // cooldown / adverse block state
    int64_t cd_until = 0;
    int    lcd = 0;      // last close direction
    int64_t lcm = 0;     // last close ms
    bool   ab = false;   // adverse block active
    double lex = 0;      // last exit price
    int    lld = 0;      // last loss direction

    int    wt = 0;       // tick count

    std::vector<Trade> trades;

    explicit Eng(const P& pp) : p(pp), ra(2.0 / (pp.rn + 1)) {}

    // mirrors ursi()
    void ursi(double mid) {
        if (rpm == 0.0) { rpm = mid; return; }
        double c = mid - rpm; rpm = mid;
        rg.push_back(c > 0 ? c : 0.0);
        rl.push_back(c < 0 ? -c : 0.0);
        int n = p.rn;
        if ((int)rg.size() > n) { rg.pop_front(); rl.pop_front(); }
        if ((int)rg.size() >= n) {
            double ag = 0, al = 0;
            for (double x : rg) ag += x;
            ag /= n;
            for (double x : rl) al += x;
            al /= n;
            rp = rc;
            rc = (al == 0.0) ? 100.0 : 100.0 - 100.0 / (1.0 + ag / al);
            double s = rc - rp;
            if (!rw) { rt_val = s; rw = true; }
            else rt_val = s * ra + rt_val * (1.0 - ra);
        }
    }

    // mirrors tick()
    void tick(int64_t ms, double bid, double ask, double drift, double atr) {
        double mid = (bid + ask) / 2.0;
        double sp  = ask - bid;
        ++wt;
        ursi(mid);
        rm_buf.push_back(mid);
        if ((int)rm_buf.size() > 10) rm_buf.pop_front();

        // drift sustained tracking
        if (drift >= p.st) {
            if (dsd != 1) { dsd = 1; dss = ms; }
        } else if (drift <= -p.st) {
            if (dsd != -1) { dsd = -1; dss = ms; }
        } else {
            dsd = 0; dss = 0;
        }
        int64_t dsms = (dsd != 0) ? (ms - dss) : 0;

        // ── manage open position ──────────────────────────────────────────
        if (pa) {
            double mv   = pl ? (mid - pe) : (pe - mid);
            if (mv > pm) pm = mv;
            double eff  = pl ? bid : ask;
            double td_  = std::fabs(ptp - pe);
            if (!pbe && td_ > 0 && mv >= td_ * 0.5) { psl = pe; pbe = true; }
            if (p.tr && mv >= p.ta) {
                double tsl = pl ? (mid - p.td) : (mid + p.td);
                if (pl  && tsl > psl) psl = tsl;
                if (!pl && tsl < psl) psl = tsl;
            }
            if ((pl && bid >= ptp) || (!pl && ask <= ptp)) { close(eff, "TP", ms); return; }
            if ((pl && bid <= psl) || (!pl && ask >= psl)) { close(eff, pbe ? "BE" : "SL", ms); return; }
            if (ms - pts > p.mh) { close(eff, "TO", ms); return; }
            return;
        }

        // ── entry gates ───────────────────────────────────────────────────
        if (wt < 300 || sp > 0.40 || atr < 1.0) return;
        if (ms < cd_until) return;
        if (!rw) return;

        // RSI direction
        int rd = 0;
        if      (rt_val >  p.rt && rt_val <  p.rm) rd =  1;
        else if (rt_val < -p.rt && rt_val > -p.rm) rd = -1;
        if (!rd) return;

        // drift agrees with RSI direction
        if (rd ==  1 && drift <  0.0) return;
        if (rd == -1 && drift >  0.0) return;
        if (std::fabs(drift) < p.dm) return;

        // sustained duration
        if (dsms < p.sm) return;

        // price confirm (last 3 ticks moving with trend)
        if ((int)rm_buf.size() >= 3) {
            double n3  = rm_buf[rm_buf.size() - 3];
            double mv3 = mid - n3;
            if (rd ==  1 && mv3 < -atr * 0.3) return;
            if (rd == -1 && mv3 >  atr * 0.3) return;
        }

        // opposite-direction cooldown (45 s)
        if (lcd != 0 && (ms - lcm) < 45000LL) {
            if (rd != lcd) return;
        }

        // adverse block
        if (ab && lld != 0) {
            double dist = (lld == 1) ? (lex - mid) : (mid - lex);
            bool same = (rd == 1 && lld == 1) || (rd == -1 && lld == -1);
            if (same && dist > atr * 0.4) return;
            else ab = false;
        }

        // ── open position ─────────────────────────────────────────────────
        bool il  = (rd == 1);
        double sl_ = atr * p.sl;
        double tp_ = sl_ * p.tp;
        double e   = il ? ask : bid;
        pa = true; pl = il; pe = e;
        psl = il ? (e - sl_) : (e + sl_);
        ptp = il ? (e + tp_) : (e - tp_);
        pts = ms; pm = 0.0; pbe = false;
        // size: matches Python ps=min(0.10,max(0.01,round(10/(sl*100)/0.01)*0.01))
        double raw = 10.0 / (sl_ * 100.0);
        ps = std::min(0.10, std::max(0.01, std::round(raw / 0.01) * 0.01));
    }

    void close(double ep, const char* reason, int64_t ms) {
        double pp = pl ? (ep - pe) : (pe - ep);
        double pu = pp * ps * 100.0;
        lcd = pl ? 1 : -1;
        lcm = ms;
        if (pu < 0) {
            ab  = true;
            lex = ep;
            lld = pl ? 1 : -1;
            cd_until = ms + p.cd;
        }
        Trade t;
        t.pnl = pu; t.mfe = pm; t.il = pl;
        t.h   = (int)((ms - pts) / 1000);
        std::strncpy(t.r, reason, 7); t.r[7] = '\0';
        trades.push_back(t);
        pa = false;
    }

    void fin(int64_t ms) { if (pa) close(pe, "END", ms); }

    double tot() const { double s = 0; for (auto& t : trades) s += t.pnl; return s; }
    double lp()  const { double s = 0; for (auto& t : trades) s += t.il ? t.pnl : 0; return s; }
    double sp2() const { double s = 0; for (auto& t : trades) s += t.il ? 0 : t.pnl; return s; }
    int    wins()const { int n = 0; for (auto& t : trades) n += (t.pnl > 0); return n; }
    int    n()   const { return (int)trades.size(); }
};

// ── Tick data ─────────────────────────────────────────────────────────────────
struct Tick { int64_t ms; double bid, ask, drift, atr; };

std::vector<Tick> load(const char* path) {
    std::vector<Tick> v;
    std::ifstream f(path);
    if (!f) { fprintf(stderr, "Cannot open %s\n", path); return v; }

    std::string line, tok;
    std::getline(f, line);

    // parse header
    int tc=-1, bc=-1, ac=-1, dc=-1, rc=-1, ci=0;
    std::istringstream hss(line);
    while (std::getline(hss, tok, ',')) {
        if (tok == "ts_ms" || tok == "timestamp_ms") tc = ci;
        if (tok == "bid")       bc = ci;
        if (tok == "ask")       ac = ci;
        if (tok == "ewm_drift") dc = ci;
        if (tok == "atr")       rc = ci;
        ++ci;
    }
    if (bc < 0 || ac < 0) { fprintf(stderr, "CSV missing bid/ask columns\n"); return v; }

    double pb = 0, av = 2.0;
    std::deque<double> aw;
    v.reserve(200000);

    while (std::getline(f, line)) {
        if (line.empty()) continue;
        std::vector<std::string> cols;
        std::istringstream ss(line);
        while (std::getline(ss, tok, ',')) cols.push_back(tok);
        int need = std::max({tc, bc, ac});
        if ((int)cols.size() <= need) continue;
        try {
            Tick t;
            t.ms    = (tc >= 0) ? (int64_t)std::stod(cols[tc]) : 0;
            t.bid   = std::stod(cols[bc]);
            t.ask   = std::stod(cols[ac]);
            t.drift = (dc >= 0 && dc < (int)cols.size()) ? std::stod(cols[dc]) : 0.0;
            t.atr   = (rc >= 0 && rc < (int)cols.size()) ? std::stod(cols[rc]) : 0.0;
            if (t.bid <= 0 || t.ask < t.bid) continue;
            // ATR fallback (mirrors Python)
            if (t.atr <= 0) {
                if (pb > 0) {
                    double tr = (t.ask - t.bid) + std::fabs(t.bid - pb);
                    aw.push_back(tr);
                    if ((int)aw.size() > 200) aw.pop_front();
                    if ((int)aw.size() >= 50) {
                        double s = 0; for (double x : aw) s += x;
                        av = s / aw.size() * 14.0;
                    }
                }
                t.atr = av;
            }
            pb = t.bid;
            v.push_back(t);
        } catch (...) {}
    }
    fprintf(stderr, "Loaded %zu ticks\n", v.size());
    return v;
}

// ── Grid (exact match to Python cfe_sweep.py) ─────────────────────────────────
// Python grid:
//   rn:  [20,30]
//   rt:  [2.0,4.0]
//   rm:  [15.0]
//   dm:  [0.3,0.5,0.8]
//   st:  [0.3,0.5]
//   sm:  [10000,20000,30000]   (ms)
//   sl:  [0.4,0.6,0.8]
//   tp:  [2.0,3.0,4.0]
//   cd:  [15000,30000]         (ms)
//   mh:  [300000,600000]       (ms)
//   tr:  [false,true]
//   ta:  [2.0]
//   td:  [1.0]
// total = 2*2*3*2*3*3*3*2*2*2 = 5184

int main(int argc, char* argv[]) {
    if (argc < 2) { puts("Usage: cfe_sweep <csv>"); return 1; }

    auto t0 = std::chrono::steady_clock::now();
    auto ticks = load(argv[1]);
    printf("Loaded %zu ticks\n", ticks.size());
    if (ticks.empty()) return 1;

    const int    rn_v[]  = {20, 30};
    const double rt_v[]  = {2.0, 4.0};
    const double rm_v[]  = {15.0};
    const double dm_v[]  = {0.3, 0.5, 0.8};
    const double st_v[]  = {0.3, 0.5};
    const int64_t sm_v[] = {10000, 20000, 30000};  // ms — matches Python
    const double sl_v[]  = {0.4, 0.6, 0.8};
    const double tp_v[]  = {2.0, 3.0, 4.0};
    const int64_t cd_v[] = {15000, 30000};          // ms — matches Python
    const int64_t mh_v[] = {300000, 600000};         // ms — matches Python
    const bool   tr_v[]  = {false, true};

    long total = 2*2*1*3*2*3*3*3*2*2*2;
    printf("Testing %ld combinations...\n", total);

    struct Res { double pnl, wr, lp, sp; int n; P p; };
    std::vector<Res> results;
    results.reserve(10000);
    long done = 0;

    for (int rn : rn_v)
    for (double rt : rt_v)
    for (double rm_ : rm_v)
    for (double dm : dm_v)
    for (double st : st_v)
    for (int64_t sm : sm_v)
    for (double sl : sl_v)
    for (double tp : tp_v)
    for (int64_t cd : cd_v)
    for (int64_t mh : mh_v)
    for (bool tr : tr_v)
    {
        P pp;
        pp.rn = rn; pp.rt = rt; pp.rm = rm_;
        pp.dm = dm; pp.st = st; pp.sm = sm;
        pp.sl = sl; pp.tp = tp; pp.cd = cd; pp.mh = mh;
        pp.tr = tr; pp.ta = 2.0; pp.td = 1.0;

        Eng eng(pp);
        for (auto& tk : ticks)
            eng.tick(tk.ms, tk.bid, tk.ask, tk.drift, tk.atr);
        eng.fin(ticks.back().ms);

        if (eng.n() >= 3) {
            Res r;
            r.pnl = eng.tot();
            r.wr  = (double)eng.wins() / eng.n();
            r.lp  = eng.lp();
            r.sp  = eng.sp2();
            r.n   = eng.n();
            r.p   = pp;
            results.push_back(r);
        }

        if (++done % 1000 == 0) {
            double best = 0;
            for (auto& r : results) if (r.pnl > best) best = r.pnl;
            printf("  %ld/%ld best=$%+.2f\n", done, total, best);
            fflush(stdout);
        }
    }

    std::sort(results.begin(), results.end(),
              [](const Res& a, const Res& b){ return a.pnl > b.pnl; });

    auto t1 = std::chrono::steady_clock::now();
    printf("\nDone in %.2fs\n",
           std::chrono::duration<double>(t1 - t0).count());

    puts("\n================================================================");
    puts("TOP 20");
    puts("================================================================");
    printf("%-3s %-8s %-6s %-4s %-7s %-7s %-5s %-5s %-5s %-4s %-5s %-3s %-5s\n",
           "#","PnL","WR","N","L$","S$","rsi_t","drift","sl","tp","sm_s","tr","hold_m");

    int show = std::min((int)results.size(), 20);
    for (int i = 0; i < show; ++i) {
        auto& r = results[i];
        printf("%-3d $%+7.2f %5.1f%% %4d $%+6.2f $%+6.2f "
               "%5.1f  %4.2f  %4.2f %4.1f  %4lds  %3s  %4ldm\n",
               i+1, r.pnl, r.wr*100, r.n, r.lp, r.sp,
               r.p.rt, r.p.dm, r.p.sl, r.p.tp,
               (long)(r.p.sm / 1000),
               r.p.tr ? "Y" : "N",
               (long)(r.p.mh / 60000));
    }

    if (!results.empty()) {
        auto& b = results[0];
        printf("\nBEST: $%+.2f  WR=%.1f%%  N=%d  L=$%+.2f  S=$%+.2f\n",
               b.pnl, b.wr*100, b.n, b.lp, b.sp);
        printf("rn=%d rt=%.1f dm=%.2f st=%.2f sm=%lds sl=%.2f tp=%.1f "
               "cd=%ldms mh=%ldms trail=%s\n",
               b.p.rn, b.p.rt, b.p.dm, b.p.st,
               (long)(b.p.sm/1000), b.p.sl, b.p.tp,
               (long)b.p.cd, (long)b.p.mh,
               b.p.tr ? "Y" : "N");

        // rerun best to show trades
        Eng eng(b.p);
        for (auto& tk : ticks)
            eng.tick(tk.ms, tk.bid, tk.ask, tk.drift, tk.atr);
        eng.fin(ticks.back().ms);

        puts("\nAll trades (best config):");
        printf("  %-5s  %-5s  %-8s  %-6s  %-4s  %s\n",
               "Side","Held","PnL","MFE","R","");
        for (auto& t : eng.trades) {
            printf("  %-5s  %4ds  $%+6.2f  %6.3f  %-4s\n",
                   t.il ? "LONG" : "SHORT",
                   t.h, t.pnl, t.mfe, t.r);
        }
    }

    return 0;
}

File: backtest/cfe_sweep.py
"""CFE Fast Sweep -- focused params only"""
import sys, csv, itertools
from collections import deque
from datetime import datetime, timezone

class CFE:
    def __init__(self, p):
        self.p=p; self.rg=deque(); self.rl=deque()
        self.rpm=0.0; self.rc=50.0; self.rp=50.0; self.rt=0.0; self.rw=False
        self.ra=2.0/(p['rn']+1); self.rm=deque()
        self.dsd=0; self.dss=0
        self.pa=False; self.pl=False; self.pe=0.0
        self.psl=0.0; self.ptp=0.0; self.pts=0; self.pm=0.0
        self.pbe=False; self.ps=0.0
        self.cd=0; self.wt=0; self.trades=[]
        self.lcd=0; self.lcm=0; self.ab=False; self.lex=0.0; self.lld=0

    def ursi(self,mid):
        if self.rpm==0: self.rpm=mid; return
        c=mid-self.rpm; self.rpm=mid
        self.rg.append(c if c>0 else 0); self.rl.append(-c if c<0 else 0)
        n=self.p['rn']
        if len(self.rg)>n: self.rg.popleft(); self.rl.popleft()
        if len(self.rg)>=n:
            ag=sum(self.rg)/n; al=sum(self.rl)/n
            self.rp=self.rc
            self.rc=100 if al==0 else 100-100/(1+ag/al)
            s=self.rc-self.rp
            if not self.rw: self.rt=s; self.rw=True
            else: self.rt=s*self.ra+self.rt*(1-self.ra)

    def tick(self,ms,bid,ask,drift,atr):
        mid=(bid+ask)/2; sp=ask-bid
        self.wt+=1; self.ursi(mid)
        self.rm.append(mid)
        if len(self.rm)>10: self.rm.popleft()
        dt=self.p['st']
        if drift>=dt:
            if self.dsd!=1: self.dsd=1; self.dss=ms
        elif drift<=-dt:
            if self.dsd!=-1: self.dsd=-1; self.dss=ms
        else:
            self.dsd=0; self.dss=0
        dsms=(ms-self.dss) if self.dsd!=0 else 0

        if self.pa:
            mv=(mid-self.pe) if self.pl else (self.pe-mid)
            if mv>self.pm: self.pm=mv
            eff=bid if self.pl else ask
            td=abs(self.ptp-self.pe)
            if not self.pbe and td>0 and mv>=td*0.5:
                self.psl=self.pe; self.pbe=True
            if self.p['tr'] and mv>=self.p['ta']:
                tsl=(mid-self.p['td']) if self.pl else (mid+self.p['td'])
                if self.pl and tsl>self.psl: self.psl=tsl
                if not self.pl and tsl<self.psl: self.psl=tsl
            if(self.pl and bid>=self.ptp)or(not self.pl and ask<=self.ptp):
                self._cl(eff,'TP',ms); return
            if(self.pl and bid<=self.psl)or(not self.pl and ask>=self.psl):
                self._cl(eff,'BE' if self.pbe else 'SL',ms); return
            if ms-self.pts>self.p['mh']: self._cl(eff,'TO',ms); return
            return

        if self.wt<300 or sp>0.40 or atr<1.0: return
        if ms<self.cd: return
        if not self.rw: return
        # RSI dir
        t=self.p['rt']; mx=self.p['rm']
        if self.rt>t and self.rt<mx: rd=1
        elif self.rt<-t and self.rt>-mx: rd=-1
        else: return
        if rd==1 and drift<0: return
        if rd==-1 and drift>0: return
        if abs(drift)<self.p['dm']: return
        if dsms<self.p['sm']: return
        # price confirm
        if len(self.rm)>=3:
            n3=list(self.rm)[-3]; mv3=mid-n3
            if rd==1 and mv3<-atr*0.3: return
            if rd==-1 and mv3>atr*0.3: return
        # opp dir cooldown
        if self.lcd!=0 and ms-self.lcm<45000:
            if (rd!=self.lcd): return
        # adverse block
        if self.ab and self.lld!=0:
            dist=(self.lex-mid) if self.lld==1 else (mid-self.lex)
            same=(rd==1 and self.lld==1)or(rd==-1 and self.lld==-1)
            if same and dist>atr*0.4: return
            else: self.ab=False
        sl=atr*self.p['sl']; tp=sl*self.p['tp']
        il=rd==1; e=ask if il else bid
        self.pa=True; self.pl=il; self.pe=e
        self.psl=e-sl if il else e+sl
        self.ptp=e+tp if il else e-tp
        self.pts=ms; self.pm=0.0; self.pbe=False
        self.ps=min(0.10,max(0.01,round(10/(sl*100)/0.01)*0.01))

    def _cl(self,ep,r,ms):
        pp=(ep-self.pe) if self.pl else (self.pe-ep)
        pu=pp*self.ps*100
        self.lcd=1 if self.pl else -1; self.lcm=ms
        if pu<0:
            self.ab=True; self.lex=ep
            self.lld=1 if self.pl else -1
            self.cd=ms+self.p['cd']
        ts=datetime.fromtimestamp(ms/1000,tz=timezone.utc).strftime('%H:%M')
        self.trades.append({'t':ts,'s':'L' if self.pl else 'S','p':pu,'m':self.pm,'r':r,'h':(ms-self.pts)//1000})
        self.pa=False

def run(f,p):
    cfe=CFE(p); aw=deque(); pb=0; av=2
    with open(f,newline='') as fh:
        for row in csv.DictReader(fh):
            try:
                ms=int(float(row.get('ts_ms',0)))
                bid=float(row['bid']); ask=float(row['ask'])
                drift=float(row.get('ewm_drift',0))
                atr=float(row.get('atr',0))
            except: continue
            if bid<=0 or ask<bid: continue
            if atr<=0:
                if pb>0:
                    aw.append(ask-bid+abs(bid-pb))
                    if len(aw)>200: aw.popleft()
                    if len(aw)>=50: av=sum(aw)/len(aw)*14
                atr=av
            pb=bid; cfe.tick(ms,bid,ask,drift,atr)
    if cfe.pa: cfe._cl(cfe.pe,'END',0)
    return cfe.trades

if len(sys.argv)<2: sys.exit(1)

# Focused grid -- parameters most likely to catch London trend moves
grid = {
    'rn':  [20,30],           # RSI period
    'ra':  [None],            # computed from rn
    'rt':  [2.0,4.0],         # RSI threshold
    'rm':  [15.0],            # RSI max
    'dm':  [0.3,0.5,0.8],     # drift min
    'st':  [0.3,0.5],         # sustained threshold
    'sm':  [10000,20000,30000],# sustained ms
    'sl':  [0.4,0.6,0.8],     # SL = sl*ATR
    'tp':  [2.0,3.0,4.0],     # TP = tp*SL
    'cd':  [15000,30000],     # cooldown ms
    'mh':  [300000,600000],   # max hold ms
    'tr':  [False,True],      # trail
    'ta':  [2.0],             # trail arm pts
    'td':  [1.0],             # trail dist pts
}

keys=list(grid.keys())
combos=list(itertools.product(*grid.values()))
print(f"Testing {len(combos)} combinations...")

results=[]
for i,combo in enumerate(combos):
    p=dict(zip(keys,combo))
    p['rn']=p['rn']; p['ra']=2.0/(p['rn']+1)  # fix alpha
    trades=run(sys.argv[1],p)
    if len(trades)>=3:
        tot=sum(t['p'] for t in trades)
        ws=[t for t in trades if t['p']>0]
        wr=len(ws)/len(trades)
        lp=sum(t['p'] for t in trades if t['s']=='L')
        sp=sum(t['p'] for t in trades if t['s']=='S')
        ln=len([t for t in trades if t['s']=='L'])
        sn=len([t for t in trades if t['s']=='S'])
        exp=tot/len(trades)
        results.append((tot,wr,len(trades),exp,lp,sp,ln,sn,p,trades))
    if (i+1)%200==0:
        best=max((r[0] for r in results),default=0)
        print(f"  {i+1}/{len(combos)}  best=${best:+.2f}")

results.sort(key=lambda x:x[0],reverse=True)

print(f"\n{'='*75}")
print(f"TOP 20 CFE CONFIGURATIONS")
print(f"{'='*75}")
print(f"{'#':3} {'PnL':8} {'WR':6} {'N':4} {'Exp':7} {'L$':7} {'S$':7} {'dm':5} {'sm':5} {'sl':5} {'tp':5} {'rt':5} {'tr'}")
for rank,(tot,wr,n,exp,lp,sp,ln,sn,p,trades) in enumerate(results[:20],1):
    print(f"{rank:3d} ${tot:+7.2f} {100*wr:5.1f}% {n:4d} ${exp:+5.3f} "
          f"${lp:+6.2f} ${sp:+6.2f} "
          f"{p['dm']:5.1f} {p['sm']//1000:4d}s "
          f"{p['sl']:5.2f} {p['tp']:5.1f} {p['rt']:5.1f} {'Y' if p['tr'] else 'N'}")

if results:
    tot,wr,n,exp,lp,sp,ln,sn,p,trades=results[0]
    print(f"\n{'='*75}")
    print(f"BEST: PnL=${tot:+.2f} WR={100*wr:.1f}% ({n} trades, {ln}L {sn}S)")
    print(f"Params: drift>={p['dm']} sus>={p['sm']//1000}s SL={p['sl']}xATR TP={p['tp']}xSL RSI_t={p['rt']} trail={'Y' if p['tr'] else 'N'} hold={p['mh']//60000}min")
    print(f"LONG: {ln} trades ${lp:+.2f}   SHORT: {sn} trades ${sp:+.2f}")
    print(f"\n  {'Time':6} {'S':2} {'R':6} {'PnL':8} {'MFE':6} {'Held'}")
    for t in trades:
        print(f"  {t['t']:6} {t['s']:2} {t['r']:6} ${t['p']:+6.2f} {t['m']:6.3f} {t['h']}s")
